docs/proposals/go-git-providers.md

```diff
@@ -0,0 +1,474 @@
+one-time initialization step.
+
+The `Repository` struct is defined as follows:
+
+```go
+// Repository represents a Git repository provided by a Git provider
+type Repository struct {
+	// RepositoryInfo provides the required fields
+	// (Provider, Domain, Organization, SubOrganizations and (Repository)Name)
+	// required for being an RepositoryRef
+	RepositoryInfo `json:",inline"`
+	// InternalHolder implements the InternalGetter interface
+	InternalHolder `json:",inline"`
+
+	// Description returns a description for the repository
+	// No default value at POST-time
+	// +optional
+	Description *string `json:"description"`
+
+	// Visibility returns the desired visibility for the repository
+	// Default value at POST-time: RepoVisibilityPrivate
+	// +optional
+	Visibility *RepoVisibility
+}
+
+// GetCloneURL gets the clone URL for the specified transport type
+func (r *Repository) GetCloneURL(transport TransportType) string {
+	return GetCloneURL(r, transport)
+}
+```
+
+As can be seen, there is also a `GetCloneURL` function for the repository which allows
+resolving the URL from which to clone the repo, for a given transport method (`ssh` and `https`
+are supported `TransportType`s)
+
+### RepositoryClient
+
+`RepositoryClient` allows access to a given repository's underlying resources, like follows:
+
+```go
+// RepositoryClient operates on a given/specific repository
+type RepositoryClient interface {
+	// Teams gives access to what teams have access to this specific repository
+	Teams() RepositoryTeamsClient
+
+	// Credentials gives access to manipulating credentials for accessing this specific repository
+	Credentials() CredentialsClient
+}
+```
+
+#### Repository Teams
+
+`RepositoryTeamsClient` allows adding & removing teams from the list of authorized persons to access a repository.
+
+```go
+// RepositoryTeamsClient operates on the teams list for a specific repository
+type RepositoryTeamsClient interface {
+	// Add adds a given team in the repo's (top-level) organization to the repository
+	Add(ctx context.Context, teamName string, opts RepositoryAddTeamOptions) error
+
+	// Remove removes the given team in the repo's (top-level) organization from the repository
+	Remove(ctx context.Context, teamName string) error
+}
+```
+
+`RepositoryAddTeamOptions` has a `Permission *TeamRepositoryPermission` which allows specifying what access level the team should have.
+
+#### Repository Credentials
+
+`CredentialsClient` allows adding & removing credentials (e.g. deploy keys) from accessing a specific repository.
+
+```go
+// CredentialsClient operates on the access credential list for a specific repository
+type CredentialsClient interface {
+	// Create a credential with the given human-readable name, the given bytes and optional options
+	Create(ctx context.Context, c RepositoryCredential, opts CredentialCreateOptions) error
+
+	// Lists all credentials
+	List(ctx context.Context) ([]RepositoryCredential, error)
```

Comment:
Do we have a practical use case of List or can we make these just `Add` and `Remove`?
If so, this List call probably should also take in a `type RepositoryCredentialType`
cc @martina-if @stefanprodan